### PR TITLE
Work-in-progress #693.

### PR DESCRIPTION
**Pull request by hzeller**
Work-in-progress: First step
  * enable the rule to allow streaming operator with 'with'
  * temporarily disable the rules that will create conflicts.
This will accept the streaming operator with 'with' syntax
at expense of other constructs.

Next step: resolve shift-reduce conflicts.

Signed-off-by: Henner Zeller <h.zeller@acm.org>

### CODE REVIEW

**Comment by hzeller** [comment]
@fangism , can you check if this approach is sound ? Description of approach in CL description of 5764608e600d2affb83f5bdc049707e43b2ff247

File: verilog/parser/verilog.lex [hunk 1]
```diff
@@ -672,6 +672,7 @@ wait_order { UpdateLocation(); return TK_wait_order; }
 wildcard { UpdateLocation(); return TK_wildcard; }
 <COVERGROUP>with { UpdateLocation(); return TK_with__covergroup; }
 with { UpdateLocation(); return TK_with; }
+"with"[ \t]*"[" { UpdateLocation(); return TK_with_followed_by_bracket; }
```
Comment (fangism) [inline verilog/parser/verilog.lex:675]: I would avoid merging this as a single token, however you can achieve context-sensitivity to give variants of the same textual token by using a combination of lexer states and `yyless` to rewind.  Keep in mind that when you create variants of tokens like `[` and `(` (and yes, even `;`) that the formatter needs to be handle these similarly.

File: verilog/parser/verilog.y [hunk 2]
```diff
@@ -3217,12 +3218,18 @@ block_item_or_statement_or_null_list_opt
...
+  | expression TK_PO_NEG expression
+  ;
+
 stream_expression
   : expression
     { $$ = move($1); }
-  /* TODO(fangism):
-  | expression TK_with select_variable_dimension
-   */
+  | expression TK_with_followed_by_bracket array_range_expression ']'
```
Comment (fangism) [inline verilog/parser/verilog.y:3232]: Related to earlier comment, separate the tokens, but you may give `[` its own flavor.

**Review by fangism — COMMENTED** [review]
General question: what was the type of conflict when you tried to uncomment the TODO regarding `with`?  Can you summarize the conflict state, with depth=1 inbound states?

You're thinking of the right kind of tactics.  Also if there's any context-sensitivity you cannot implement in the lexer alone, there is lexical_context.{h,cc} available.

**Comment by hzeller** [comment]
> General question: what was the type of conflict when you tried to uncomment the TODO regarding `with`? Can you summarize the conflict state, with depth=1 inbound states?

Not entirely sure what you mean that goes beyond the change description in 5764608e600d2affb83f5bdc049707e43b2ff247 (is there a particular output that I should be looking at from bison and extract that from there ?), but the problem arose when seeing `with` in this context:

```yacc
stream_expression
   : expression TK_with '[' array_range_expression ']'
```

Then the parser could shift assuming that this is an expression that has an optional `with` at the end (like expressions with constraint block), or reduce the previous expression and move on to the stream expression with-clause. So only seeing one token ahead (`TK_with`), created this conflict because only then the next token would decide what that is (which could either be `(` or `[`).

Solved this by inspecting that all the other cases are followed with something not `[`, so this could be disambiguated by just peeking ahead in the lexer.

File: verilog/parser/verilog.lex (hunk 1 above)
Comment (hzeller) [reply verilog/parser/verilog.lex:675]: Thanks, I think I'll reformulate it like that.

Comment (fangism) [reply verilog/parser/verilog.lex:675]: Another thing to keep in mind is *comments*.  You'll notice that in various state blocks in the lexer, I had to explicitly add comment handling to push state.

**Comment by fangism** [comment]
> > General question: what was the type of conflict when you tried to uncomment the TODO regarding `with`? Can you summarize the conflict state, with depth=1 inbound states?
> 
> Not entirely sure what you mean that goes beyond the change description in [5764608](https://github.com/google/verible/commit/5764608e600d2affb83f5bdc049707e43b2ff247) (is there a particular output that I should be looking at from bison and extract that from there ?), but the problem arose when seeing `with` in this context:
> 
> ```yacc
> stream_expression
>    : expression TK_with '[' array_range_expression ']'
> ```
> 
> Then the parser could shift assuming that this is an expression that has an optional `with` at the end (like expressions with constraint block), or reduce the previous expression and move on to the stream expression with-clause. So only seeing one token ahead (`TK_with`), created this conflict because only then the next token would decide what that is (which could either be `(` or `[`).
> 
> Solved this by inspecting that all the other cases are followed with something not `[`, so this could be disambiguated by just peeking ahead in the lexer.

If one token look-ahead is enough then this might work.  If it handling comments (e.g. after `TK_with`) directly in the lexer starts to get messy, this might be done as simply in the lexical context class, where you'd add a tiny-state machine that ignores comments.

**Comment by hzeller** [comment]
mmh, I rebased the branch, but that looks like it created a mess.
I'll create a separate pull request.
